src/gt4py/stencil_object.py

```diff
@@ -267,6 +297,15 @@ def _validate_args(self, field_args, param_args, domain, origin) -> None:
                         f"expects {field_domain_ndim + len(field_info.data_dims)} ('{field_info.axes}[{field_info.data_dims}]')"
                     )
 
+                if (
+                    isinstance(field, gt_storage.storage.Storage)
+                    and tuple(field.mask)[:domain_ndim] != field_domain_mask
+                ):
+                    raise ValueError(
+                        f"Storage for '{name}' has domain mask '{field.mask}' but the API signature "
+                        f"expects '[{', '.join(field_info.axes)}]'"
+                    )
+
```

Comment:
Are there tests to exercise this new condition? Isn't it redundant after the checks in `# assert compatibility of fields with stencil` (line 253 ...)?

Reply:
The test was moved over from `_get_max_domain`, but you're right that I don't think both that and the test for number of dimensions is needed.
```python
                if field.ndim != field_domain_ndim + len(field_info.data_dims):
                    raise ValueError(
                        f"Storage for '{name}' has {field.ndim} dimensions but the API signature "
                        f"expects {field_domain_ndim + len(field_info.data_dims)} ('{field_info.axes}[{field_info.data_dims}]')"
                    )

                if (
                    isinstance(field, gt_storage.storage.Storage)
                    and tuple(field.mask)[:domain_ndim] != field_domain_mask
                ):
                    raise ValueError(
                        f"Storage for '{name}' has domain mask '{field.mask}' but the API signature "
                        f"expects '[{', '.join(field_info.axes)}]'"
                    )
```
I suggest we remove the one for the number of dimensions. That causes one test in `test_call_interface` to fail, but it triggers on the other exception so I can fix it.

Reply:
If `in_storage = gt_storage.ones(shape=(4, 4, 2), backend="gtc:numpy", default_origin=(0, 0, 0), dtype=(np.float_, (3, 3)))` then the dimension exception is raised.

Reply:
I opened up #542 to address this.

Reply:
Ok, it makes sense now. Thanks for the clarification.